### PR TITLE
Update sparse_embed.py

### PR DESCRIPTION
**Pull request by arthur-75**
Hello, we need all the indices to be compare and not only the one is scored by SPLADE, as SPLADE is multiplying the logits and reranking them by logits where the ranking of SparsEmbed is based only on Contextual Embeddings.

### CODE REVIEW

File: neural_cherche/retrieve/bm25.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    ) -> "CountVectorizer":
+        """Add new documents to the CountVectorizer retriever."""
+        for document_key in documents_embeddings:
+            self.documents.append({self.key: document_key})
+
+        self.n_documents += len(documents_embeddings)
+        n_samples,n_features  = self.matrix.shape
+        df = np.bincount(self.matrix.indices, minlength=n_features)#Count the number of non-zero values for each feature in sparse X
+        idf = np.log(n_samples / (df)) #compute the log of idf
+        avdl  = self.matrix.sum(1).mean() # mean of all
```
Comment (raphaelsty) [inline neural_cherche/retrieve/bm25.py:123]: Would it make sense to call self.matrix.sum(1) once and then call .mean() and .A1 to get avdl and len_X ? If self.matrix.sum(1) does not cost anything to compute we can keep it like this

File: neural_cherche/retrieve/bm25.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+            
+            batch_match, batch_similarities = self.top_k(batch_queries=batch_queries, k=k)
+
+            for match, similarities in zip(batch_match, batch_similarities):
+                
+                ranked.append(
+                    [
+                        {**self.documents[idx], "similarity": similarity}
+                        for idx, similarity in zip(match, similarities)
+                        #if similarity > 0
```
Comment (raphaelsty) [inline neural_cherche/retrieve/bm25.py:186]: We did filter similarity > 0 in order to only retrieve top_k documents that match the query. Does the function top_k only return documents with similarity > 0 ? Otherwise I think we should use this condition

File: neural_cherche/retrieve/sparse_embed.py [hunk 3]
```diff
@@ -330,8 +330,10 @@ def _retrieve(
         k
             Number of documents to retrieve.
         """
-        similarities = -1 * embeddings["sparse_activations"].dot(self.matrix)
-        sparse_matchs, _ = self.top_k(similarities=similarities, k=k)
+        #similarities = -1 * embeddings["sparse_activations"].dot(self.matrix)
+        #sparse_matchs, _ = self.top_k(similarities=similarities, k=k)
+        sparse_matchs=[ i.indices for i in  (embeddings["sparse_activations"].dot(self.matrix))]
```
Comment (raphaelsty) [inline neural_cherche/retrieve/sparse_embed.py:335]: The contextual scoring function of SparseEmbed is quite costly, do you think it's better to not filter that match the sparse scoring function in order to rerank top_k documents?

File: neural_cherche/retrieve/bm25.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    2. [Python: tf-idf-cosine: to find document similarity](https://stackoverflow.com/questions/12118720/python-tf-idf-cosine-to-find-document-similarity)
+
+    """
+    def __init__(
+        self,
+        key: str,
+        on: list[str],
+        b=0.75, 
+        k1=1.5,
+        countVector=None,
```
Comment (raphaelsty) [inline neural_cherche/retrieve/bm25.py:78]: We should replace `countVector` by `count_vectorizer` in order to match snake case convention

File: neural_cherche/retrieve/bm25.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        self.key = key
+        self.on = [on] if isinstance(on, str) else on
+        self.countVector=(
+            CountVectorizer(lowercase=True,ngram_range=(3, 7), analyzer="char")
+            if countVector is None
+            else countVector
+        )
+        #Tunable parameters
+        self.b = b # The impact of document length normalization, Higher --> penalize longer documents more.
+        self.k1 = k1 #how quickly the impact of term frequency saturates, Higher --> make term frequency more influential
```
Comment (raphaelsty) [inline neural_cherche/retrieve/bm25.py:89]: You can remove the comments here and rather add comments in the docstring. Those comments will be parsed and used by the documentation

File: neural_cherche/retrieve/bm25.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    ----------
+    1. [sklearn.feature_extraction.text.CountVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html)
+    2. [Python: tf-idf-cosine: to find document similarity](https://stackoverflow.com/questions/12118720/python-tf-idf-cosine-to-find-document-similarity)
+
+    """
+    def __init__(
+        self,
+        key: str,
+        on: list[str],
+        b=0.75,
```
Comment (raphaelsty) [inline neural_cherche/retrieve/bm25.py:76]: You can add type hinting here, b: float = 0.75, k: float = 1.5, count_vectorizer: CountVectorizer

**Review by raphaelsty — COMMENTED** [review]
This is a great improvement to neural-cherche ! 🥳

I made a few comments that are mostly related to code style. It would be great to run [black](https://black.readthedocs.io/en/stable/getting_started.html) and [isort](https://pycqa.github.io/isort/) over your code. 

Kudos @arthur-75

File: neural_cherche/retrieve/bm25.py (hunk 2 above)
Comment (arthur-75) [reply neural_cherche/retrieve/bm25.py:186]: I was testing something I will uncomment it back

File: neural_cherche/retrieve/sparse_embed.py (hunk 3 above)
Comment (raphaelsty) [reply neural_cherche/retrieve/sparse_embed.py:335]: Hi @arthur-75, kudos for the update of the pull-request, you did not respond to this specific comment, maybe because you did not send your comments ? You can use "add single comment" button in order to send your feedback otherwise you can use "start a review" and then press another button to validate the review. 

image
